### PR TITLE
[NFC][SYCL] Switch to std::enable_if

### PR DESCRIPTION
**Pull request by aelovikov-intel**

### CODE REVIEW

File: sycl/include/sycl/detail/stl_type_traits.hpp [hunk 1]
```diff
@@ -20,9 +20,6 @@ namespace detail {
 // Type traits identical to those in std in newer versions. Can be removed when
 // SYCL requires a newer version of the C++ standard.
 // C++14
-template <bool B, class T = void>
```
Comment (aelovikov-intel) [inline sycl/include/sycl/detail/stl_type_traits.hpp:23]: `sycl::detail::enable_if` is removed here.

**Comment by bader** [comment]
No "#include <sycl/include/sycl/detail/stl_type_traits.hpp>" -> "#include <type_traits>"? I suppose this type of transformation is not automated. Right?

**Comment by aelovikov-intel** [comment]
> No "#include <sycl/include/sycl/detail/stl_type_traits.hpp>" -> "#include <type_traits>"? I suppose this type of transformation is not automated. Right?

I plan to switch different utilities one by one. Once all uses are cleaned up, I'll remove the header.

**Comment by bader** [comment]
> > No "#include <sycl/include/sycl/detail/stl_type_traits.hpp>" -> "#include <type_traits>"?
> 
> I plan to switch different utilities one by one. Once all uses are cleaned up, I'll remove the header.
> 
> > I suppose this type of transformation is not automated. Right?
> 
> I don't understand the question. What I did is
> 
> 1. Emacs' M-x rgrep enable_if
> 2. Switch to grep-writable-mode, perform regexp, save
> 3. Run clang-format

Right. I assume that some files are using only `enable_if` utility. If so, we can include the right header for these files right away. The problem is you need check if any other utility functions are not used. My question was about automating this task with tools like [include-what-you-use](https://github.com/include-what-you-use/include-what-you-use). If we can't update includes automatically, I don't think it's worth doing manually.

**Comment by aelovikov-intel** [comment]
> If we can't update includes automatically, I don't think it's worth doing manually.

I don't have experience with it and setting it up might take more time than removing the header completely :)

**Review by kbobrovs — APPROVED** [review]
ESIMD LGTM

**Review by v-klochkov — APPROVED** [review]

**Comment by againull** [comment]
Could you please fix RHEL build failure, seems relate: http://icl-jenkins2.sc.intel.com:8080/blue/organizations/jenkins/SYCL_CI%2Fintel%2FBuild_PR_RHEL/detail/Build_PR_RHEL/3931/pipeline/152

**Review by pvchupin — APPROVED** [review]
cmake changes looks ok

**Comment by hdelan** [comment]
This PR has introduced a build failure due to missing includes, ping @bader 
```
[148/220] Generating ../../lib/libsycl-fallback-complex.spv
FAILED: lib/libsycl-fallback-complex.spv /home/hugh/llvm/build/lib/libsycl-fallback-complex.spv 
cd /home/hugh/llvm/build/tools/libdevice && /home/hugh/llvm/build/bin/clang-16 -fsycl-device-only -fno-sycl-use-bitcode -Wno-sycl-strict -Wno-undefined-internal -sycl-std=2020 --gcc-toolchain=/usr/lib -fno-sycl-libspirv -fno-bundle-offload-arch -nocudalib /home/hugh/llvm/libdevice/fallback-complex.cpp -o /home/hugh/llvm/build/./lib/libsycl-fallback-complex.spv
clang-16: warning: ignoring '-fno-sycl-libspirv' option as it is not currently supported for target 'spir64-unknown-unknown' [-Woption-ignored]
clang-16: warning: argument unused during compilation: '-fno-bundle-offload-arch' [-Wunused-command-line-argument]
/home/hugh/llvm/libdevice/fallback-complex.cpp:12:10: fatal error: 'cmath' file not found
#include <cmath>
         ^~~~~~~
1 error generated.
```

File: libdevice/cmake/modules/SYCLLibdevice.cmake [hunk 2]
```diff
@@ -32,6 +32,15 @@ set(compile_opts
   -sycl-std=2020
   )
 
+if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
```
Comment (hdelan) [inline libdevice/cmake/modules/SYCLLibdevice.cmake:35]: What was the reason for this change @aelovikov-intel ?

Comment (hdelan) [reply libdevice/cmake/modules/SYCLLibdevice.cmake:35]: https://github.com/intel/llvm/pull/7761/

Comment (aelovikov-intel) [reply libdevice/cmake/modules/SYCLLibdevice.cmake:35]: Our testing on a system with older libc where gcc installation is in non-default location (as default is too old).

**Pull request merged**